aip/0146.md

```diff
@@ -0,0 +1,96 @@
+
+However, occasionally it is appropriate to have a generic or polymorphic field
+of some kind that can conform to multiple schemata, or even be entirely
+free-form.
+
+## Guidance
+
+While generic fields are generally rare, a service **may** introduce generic
+field where necessary. There are several approaches to this depending on how
+generic the field needs to be; in general, services **should** attempt to
+introduce the "least generic" approach that is able to satisfy the use case.
+
+### Oneof
+
+A `oneof` **may** be used to introduce a type union: the user or service is
+able to specify one of the fields inside the oneof. Additionally, a `oneof`
+**may** be used with the same type (usually strings) to represent a semantic
+difference between the options.
+
+Because the individual fields in the oneof have different keys, a developer can
+programmatically determine which (if any) of the fields is populated.
+
+A `oneof` preserves the largest degree of type safety and semantic meaning for
+each option, and services **should** generally prefer them over other generic
+or polymorphic options when feasible. However, the `oneof` construct is
+ill-suited when there is a large (or unlimited) number of potential options, or
+when there is a large resource structure that would require a long series of
+"cascading oneofs".
+
+**Note:** Adding additional possible fields to an existing `oneof` is a
+non-breaking change, but moving existing fields into or out of a `oneof` is
+breaking (it creates a backwards-incompatible change in Go protobuf stubs).
+
+### Maps
+
+Maps **may** be used in situations where a situation where many values _of the
+same type_ are needed, but the keys are unknown or user-determined.
+
+Maps are usually not appropriate for generic fields because the map values all
+share a type, but occasionally they are useful. In particular, a map can
+sometimes be suited to a situation where many objects of the same type are
+needed, with different behavior based on the names of their keys (for example,
+using keys as environment names).
+
+### Struct
+
+The [`google.protobuf.Struct`][] object **may** be used to represent arbitrary
+nested JSON. Keys can be strings, and values can be floats, strings, booleans,
+arrays, or additional nested structs, allowing for an arbitrarily nested
+structure that can be represented as JSON (and is automatically represented as
+JSON when using REST/JSON).
+
+A `Struct` is most useful when the service does not know the schema in advance,
+or when a service needs to store and retrieve arbitrary but structured user
+data. Using a `Struct` is convenient for users in this case because they can
+easily get JSON objects that can be natively manipulated in their environment
+of choice.
+
+If a service needs to reason about the _schema_ of a `Struct`, it **should**
+use [JSONSchema][] for this purpose. Because JSONSchema is itself JSON, a valid
+JSONSchema document can itself be stored in a `Struct`.
```

Comment:
I don't believe we have any guidance on that, no.

Reply:
Ah, ok thanks for getting back to me. Are you aware of any API's that you've used that comply with this recommendation?

Reply:
or @apasel422 ?

Reply:
Not offhand, I'm afraid.

Reply:
Struct was designed to pass JSON objects around without data loss (except you need to pass larger JSON numbers as strings in Struct). JSON doesn't have built-in schema, neither Struct.

You can use external systems to define JSON schemas. One popular solution is OpenAPI spec. Google APIs use proto3 to define all schemas, see github.com/googleapis/googleapis. The compiled schema is known as a service config, `google.api.Service`. This format is not widely used outside Google. Google API Discovery Document is derived from the service config, and used for generating client libraries. What developers actually use (90% of usage) is using the client libraries, and the schema is presented as Java objects etc. Protobuf Descriptor can be another format.

Since AIP is about APIs, I would recommend to use OpenAPI spec. How to easily generate the OpenAPI spec is a separate question. I don't know if there is a free and high-quality tool for it. Google used to provide a tool called API Compiler, but it was only an experiment and withdrawn years ago.

Reply:
Thanks @wora !